apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/SamplingProfiler.java

```diff
@@ -0,0 +1,287 @@
+
+    void setProfilingSessionOngoing(boolean profilingSessionOngoing) {
+        this.profilingSessionOngoing = profilingSessionOngoing;
+    }
+
+    void clear() {
+        profiledThreads.clear();
+        activationEvents.clear();
+    }
+
+    @Override
+    public void run() {
+        TimeDuration sampleRate = config.getSampleRate();
+        TimeDuration profilingDuration = config.getProfilingDuration();
+        TimeDuration profilingInterval = config.getProfilingInterval();
+        profilingSessionOngoing = true;
+
+        logger.debug("Start profiling session");
+        profile(sampleRate, profilingDuration);
+        logger.debug("End profiling session");
+
+        if (profilingInterval.getMillis() != 0) {
+            profilingSessionOngoing = false;
+            // TODO do we want to create inferred spans for partially profiled transactions?
+            profiledThreads.clear();
+        }
+        // clears the interrupted status so that the thread can return to the pool
+        if (!Thread.interrupted()) {
+            long delay = profilingInterval.getMillis() - config.getProfilingDuration().getMillis();
+            scheduler.schedule(this, delay, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private void profile(TimeDuration sampleRate, TimeDuration profilingDuration) {
+        long sampleRateMs = sampleRate.getMillis();
+        long deadline = profilingDuration.getMillis() + System.currentTimeMillis();
+        while (!Thread.currentThread().isInterrupted() && System.currentTimeMillis() < deadline) {
+            try {
+                long startNs = System.nanoTime();
+
+                takeThreadSnapshot();
+
+                long durationMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);
+                logger.trace("Taking snapshot took {}ms", durationMs);
+
+                Thread.sleep(Math.max(sampleRateMs - durationMs, 0));
+            } catch (RuntimeException e) {
+                logger.error("Exception while taking profiling snapshot", e);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                break;
+            }
+        }
+    }
+
+    private void takeThreadSnapshot() {
+        List<WildcardMatcher> excludedClasses = config.getExcludedClasses();
+        List<WildcardMatcher> includedClasses = config.getIncludedClasses();
+        long nanoTime = System.nanoTime();
+        processActivationEventsUpTo(nanoTime);
+        long[] profiledThreadIds = getProfiledThreadIds();
+        ThreadInfo[] threadInfos = threadMXBean.getThreadInfo(profiledThreadIds, Integer.MAX_VALUE);
+        if (profiledThreadIds.length == 0) {
+            return;
+        }
+        logger.trace("Taking snapshot of threads {} timestamp {}", profiledThreadIds, nanoTime);
+
+        List<StackFrame> stackTraces = new ArrayList<>(256);
+        for (int i = 0; i < profiledThreadIds.length; i++) {
+            long threadId = profiledThreadIds[i];
```

Comment:
So here, if active span is not the same as you saw before taking the stack trace- discard this stack trace.
You just need to make sure the Root#activeSpan is volatile

Reply:
Updating the `CallTree` and taking thread dumps is deliberately done on the same thread so that there can't be any concurrency issues. There's a mechanism to process `ActivationEvents` up to a certain nano timestamp from the ring buffer. So there can't be any race conditions from concurrent activations as they are just put in the queue but they don't modify the `CallTree`.